python/dawn/sir.py

```diff
@@ -0,0 +1,594 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+##===-----------------------------------------------------------------------------*- Python -*-===##
+##                          _                      
+##                         | |                     
+##                       __| | __ ___      ___ ___  
+##                      / _` |/ _` \ \ /\ / / '_  | 
+##                     | (_| | (_| |\ V  V /| | | |
+##                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+##
+##
+##  This file is distributed under the MIT License (MIT). 
+##  See LICENSE.txt for details.
+##
+##===------------------------------------------------------------------------------------------===##
+
+from collections import Iterable
+from sys import path as sys_path
+from typing import List, TypeVar
+
+from dawn.config import __dawn_install_protobuf_module__
+
+from dawn.error import ParseError, SIRError
+
+sys_path.insert(1, __dawn_install_protobuf_module__)
+
+#
+# Export all SIR classes
+#
+from .SIR_pb2 import *
+from google.protobuf import json_format
+
+ExprType = TypeVar('Expr',
+                   Expr,
+                   UnaryOperator,
+                   BinaryOperator,
+                   AssignmentExpr,
+                   TernaryOperator,
+                   FunCallExpr,
+                   StencilFunCallExpr,
+                   StencilFunArgExpr,
+                   VarAccessExpr,
+                   FieldAccessExpr,
+                   LiteralAccessExpr,
+                   )
+
+StmtType = TypeVar('Stmt',
+                   Stmt,
+                   BlockStmt,
+                   ExprStmt,
+                   ReturnStmt,
+                   VarDeclStmt,
+                   VerticalRegionDeclStmt,
+                   StencilCallDeclStmt,
+                   BoundaryConditionDeclStmt,
+                   IfStmt
+                   )
+
+
+def to_json(msg):
+    """ Converts protobuf message to JSON format.
+
+    :param msg: The protocol buffers message instance to serialize.
+    :returns: A string containing the JSON formatted protocol buffer message.
+    """
+    return json_format.MessageToJson(msg)
+
+
+def from_json(text: str, message_type):
+    """ Parses a JSON representation of a protocol message and returns the parsed message.
+
+    :param text: Text JSON representation of the message.
+    :param message_type: The *type* of message to parse.
+    :returns: The parsed message.
+    :raises ParseError: Failed to parse JSON
+    """
+    msg = message_type()
+    try:
+        json_format.Parse(text, msg)
+    except json_format.ParseError as e:
+        raise ParseError(str(e))
+    return msg
+
+
+def makeType(builtin_type_or_name, is_const: bool = False, is_volatile: bool = False):
```

Comment:
What is the convention for function annotations here? Are all the function-annotated functions exposed in an interface to C?

Reply:
Functions annotations are just for static type checking by linter tools and IDEs (it is completely optional in Python)